algo/src/main/java/org/neo4j/graphalgo/walking/NodeWalkerProc.java

```diff
@@ -0,0 +1,167 @@
+
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.stream.*;
+
+public class NodeWalkerProc  {
+
+    private static final RelationshipType NEXT = RelationshipType.withName("NEXT");
+    @Context
+    public GraphDatabaseAPI api;
+
+    @Context
+    public Log log;
+
+    @Context
+    public KernelTransaction transaction;
+
+
+
+    @Procedure(name = "algo.randomWalk.stream", mode = Mode.READ)
+    @Description("CALL algo.randomWalk.stream(label, relationship, start:null=all/[ids]/label, steps, walks, {mode:random/node2vec, return:1.0, inOut:1.0, path:false/true concurrency:4, direction:'BOTH'}) " +
+            "YIELD nodes, path - computes random walks from given starting points")
+    public Stream<WalkResult> randomWalk(
+            @Name(value = "label", defaultValue = "") String label,
+            @Name(value = "relationship", defaultValue = "") String relationship,
+            @Name(value = "start", defaultValue = "null") Object start,
+            @Name(value = "steps", defaultValue = "10") long steps,
+            @Name(value = "walks", defaultValue = "1") long walks,
+            @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+
+        ProcedureConfiguration configuration = ProcedureConfiguration.create(config);
+
+        PageRankScore.Stats.Builder statsBuilder = new PageRankScore.Stats.Builder();
+
+        AllocationTracker tracker = AllocationTracker.create();
+
+        Direction direction = configuration.getDirection(Direction.BOTH);
+
+        final Graph graph = load(label, relationship, tracker, configuration.getGraphImpl(), statsBuilder, configuration);
+
+        int nodeCount = Math.toIntExact(graph.nodeCount());
+
+        if(nodeCount == 0) {
+            graph.release();
+            return Stream.empty();
+        }
+
+
+        Number returnParam = configuration.get("return", 1d);
+        Number inOut = configuration.get("inOut", 1d);
+        NodeWalker.NextNodeStrategy strategy = configuration.get("mode","random").equalsIgnoreCase("random") ?
+                new NodeWalker.RandomNextNodeStrategy(graph, graph) :
+                new NodeWalker.Node2VecStrategy(graph,graph, returnParam.doubleValue(), inOut.doubleValue());
+
+        TerminationFlag terminationFlag = TerminationFlag.wrap(transaction);
+
+        int concurrency = configuration.getConcurrency();
+
+        Boolean returnPath = configuration.get("path", false);
+
+        int limit = (walks == -1) ? nodeCount : Math.toIntExact(walks);
+
+        PrimitiveIterator.OfInt idStream = IntStream.range(0, limit).unordered().parallel().flatMap((s) -> idStream(start, graph, limit)).limit(limit).iterator();
+
+        Stream<long[]> randomWalks = new NodeWalker().randomWalk(graph, (int) steps, strategy, terminationFlag, concurrency, limit, idStream);
+        return randomWalks
+                .map( nodes -> new WalkResult(nodes, returnPath ? toPath(api, nodes,direction) : null));
```

Comment:
Should we return the nodeId where the random walk begins from as part of the stream as well, or should users just assume that is the first element of the `nodes` array?

Reply:
That was at least what the students used, and also in a path it's always the first node-id.
You definitely need to return it as most uses don't have a single start id.
But we can also add an extra column if that makes sense. I wouldn't want to remove it from the array though as that breaks the path.